pkg/logx/middleware.go

```diff
@@ -0,0 +1,200 @@
+	AfterNextSkipper middleware.Skipper
+	// BeforeNext is a function that is executed before the next handler is called
+	BeforeNext middleware.BeforeFunc
+	// Enricher is a function that can be used to enrich the logger with additional information
+	Enricher Enricher
+	// RequestIDHeader is the header name to use for the request ID in a log record
+	RequestIDHeader string
+	// RequestIDKey is the key name to use for the request ID in a log record
+	RequestIDKey string
+	// NestKey is the key name to use for the nested logger in a log record
+	NestKey string
+	// HandleError indicates whether to propagate errors up the middleware chain, so the global error handler can decide appropriate status code
+	HandleError bool
+	// For long-running requests that take longer than this limit, log at a different level
+	RequestLatencyLimit time.Duration
+	// The level to log at if RequestLatencyLimit is exceeded
+	RequestLatencyLevel zerolog.Level
+}
+
+// Enricher is a function that can be used to enrich the logger with additional information
+type Enricher func(c echo.Context, logger zerolog.Context) zerolog.Context
+
+// Context is a wrapper around echo.Context that provides a logger
+type Context struct {
+	echo.Context
+	logger *Logger
+}
+
+// NewContext returns a new Context
+func NewContext(ctx echo.Context, logger *Logger) *Context {
+	return &Context{ctx, logger}
+}
+
+func (c *Context) Logger() echo.Logger {
+	return c.logger
+}
+
+// LoggingMiddleware is a middleware that logs requests using the provided logger
+func LoggingMiddleware(config Config) echo.MiddlewareFunc {
+	if config.Skipper == nil {
+		config.Skipper = middleware.DefaultSkipper
+	}
+
+	if config.AfterNextSkipper == nil {
+		config.AfterNextSkipper = middleware.DefaultSkipper
+	}
+
+	if config.Logger == nil {
+		config.Logger = New(os.Stdout, WithTimestamp())
+	}
+
+	if config.RequestIDKey == "" {
+		config.RequestIDKey = "request_id"
+	}
+
+	if config.RequestIDHeader == "" {
+		config.RequestIDHeader = "X-Request-ID"
+	}
+
+	return func(next echo.HandlerFunc) echo.HandlerFunc {
+		return func(c echo.Context) error {
+			if config.Skipper(c) {
+				return next(c)
+			}
+
+			var err error
+
+			req := c.Request()
+			start := time.Now()
+
+			logger := config.Logger
+
+			logger = enrichLogger(c, logger, config)
+
+			id := getRequestID(c, config)
+			if id != "" {
+				logger = From(logger.log, WithField(config.RequestIDKey, id))
+			}
+
+			// The request context is retrieved and set to the logger's context
+			// the context is then set to the request, and a new context is created with the logger
+			c.SetRequest(req.WithContext(logger.WithContext(req.Context())))
+			c = NewContext(c, logger)
+
+			if config.BeforeNext != nil {
+				config.BeforeNext(c)
+			}
+
+			if err = next(c); err != nil {
+				if config.HandleError {
+					c.Error(err)
+				}
+			}
+
+			if config.AfterNextSkipper(c) {
+				return err
```

Comment:
Theres no `err` here so its always nil; we should just return that instead
```suggestion
				return nil
```